internal/component/discovery/distributed_targets.go

```diff
@@ -0,0 +1,91 @@
+package discovery
+
+import (
+	"github.com/grafana/ckit/peer"
+	"github.com/grafana/ckit/shard"
+
+	"github.com/grafana/alloy/internal/service/cluster"
+)
+
+// DistributedTargets uses the node's Lookup method to distribute discovery
+// targets when a component runs in a cluster.
+type DistributedTargets struct {
+	localTargets []Target
+	// localTargetKeys is used to cache the key hash computation. Improves time performance by ~20%.
+	localTargetKeys  []shard.Key
+	remoteTargetKeys map[shard.Key]struct{}
+}
+
+// NewDistributedTargets creates the abstraction that allows components to
+// dynamically shard targets between components.
+func NewDistributedTargets(clusteringEnabled bool, cluster cluster.Cluster, allTargets []Target) *DistributedTargets {
+	if !clusteringEnabled || cluster == nil {
+		cluster = disabledCluster{}
+	}
+
+	localCap := len(allTargets) + 1
+	if peerCount := len(cluster.Peers()); peerCount != 0 {
+		localCap = (len(allTargets) + 1) / peerCount
+	}
+
+	localTargets := make([]Target, 0, localCap)
+	localTargetKeys := make([]shard.Key, 0, localCap)
+	remoteTargetKeys := make(map[shard.Key]struct{}, len(allTargets)-localCap)
+
+	for _, tgt := range allTargets {
+		targetKey := keyFor(tgt)
+		peers, err := cluster.Lookup(targetKey, 1, shard.OpReadWrite)
+		belongsToLocal := err != nil || len(peers) == 0 || peers[0].Self
+
+		if belongsToLocal {
+			localTargets = append(localTargets, tgt)
+			localTargetKeys = append(localTargetKeys, targetKey)
+		} else {
+			remoteTargetKeys[targetKey] = struct{}{}
+		}
+	}
+
+	return &DistributedTargets{
+		localTargets:     localTargets,
+		localTargetKeys:  localTargetKeys,
+		remoteTargetKeys: remoteTargetKeys,
+	}
+}
+
+// LocalTargets returns the targets that belong to the local cluster node.
+func (dt *DistributedTargets) LocalTargets() []Target {
+	return dt.localTargets
+}
+
+// MovedToRemoteInstance returns targets that have been moved from this local node to another node, given the provided
+// previous targets distribution. If a target has simply disappeared, it will not be returned. Previous targets
+// distribution can be nil, in which case no targets will be returned.
+func (dt *DistributedTargets) MovedToRemoteInstance(prev *DistributedTargets) []Target {
```

Comment:
It took me a little bit of time to wrap my head around whether this comment aligns with the implementation. I think it might be worth making this a little more explicit like:

```
// MovedToRemoteInstance returns the set of local targets from prev 
// that are no longer local in dt, indicating an active target has moved. 
// Only targets which exist in both prev and dt are returned. If prev 
// contains an empty list of targets, no targets are returned. 
```

Personally I think mentioning the parameter and receiver names makes it a bit easier for me to understand, though I have a hard time understanding why.